Fix external apps returning data when Collect is being restored

seadowg opened the pull request:
Closes #5646
Potentially fixes [this crash](https://console.firebase.google.com/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/e3a3fb3d7d9a95533db04ddc128acc25?time=last-seven-days&versions=v2023.2.3%20(4681)&types=crash&sessionEventKey=64D0A9E70053000112F56C3F4D67F098_1842938818000071237)

As well as fixing the issue above, this fixes the problem discovered by @grzesiek2010:

> 1. Set "Background process limit" to "No background processes"
> 2. Open a form with an image widget
> 3. Click "Take picture"
> 4. Switch to another app and get back to the camera app
> 5. Take a picture in the camera app

#### What has been done to verify that this works as intended?

New (local) tests written to both replace and extend (the now deleted) `ProcessRestoreTest`. These tests use raw Robolectric rather than AndroidX as testing lifecycles like this requires features that [AndroidX does not support](https://github.com/android/android-test/issues/1825). As part of writing these tests, I've pulled a few things out of `androidTest` to shared modules (such as `testshared` and `androidtest` for example) so as not to duplicate test support code in `test`. I also moved our test forms to a new module that allows local tests to access them (also to prevent duplication).

#### Why is this the best possible solution? Were any other approaches considered?

I've just made the minimum changes required to get the new tests passing. I think there's a lot we could do to make this nicer using the new save instance state/activity result APIs, removing reliance on static state etc, but most of the effort here was getting reliable testing in place which I think has given us a big enough PR.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

As well as the issues this fixes, I think testing form loading/entry in general and having a play with "Don't keep activities" and "Background process limit" would be good.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/FormFillingActivity.java:
```diff
@@ -2098,17 +2092,6 @@ public void loadingComplete(FormLoaderTask task, FormDef formDef, String warning
 
-            boolean pendingActivityResult = task.hasPendingActivityResult();
-
-            if (pendingActivityResult) {
```
I believe this mechanism was used to handle receiving activity results like for example images (ImageWidget) after process death?
Now (even without this change but using the current master) if you have no background processes set in `Developer options` the result (image in this case) is ignored and the form is reloaded which means we end up in the hierarchy view. As I remember in the past it was possible to receive activity results in such a case. The changes you have implemented in that area recently changed that and I don't know if that was a deliberate decision.
I'm afraid that when using some older devices with little memory it might be difficult to receive some activity results when switching to other apps is required like when taking photos because then the form is reloaded and we are moved to the hierarchy view and the result is ignored.

seadowg replied:
> Now (even without this change but using the current master) if you have no background processes set in Developer options the result (image in this case) is ignored and the form is reloaded which means we end up in the hierarchy view. As I remember in the past it was possible to receive activity results in such a case. The changes you have implemented in that area recently changed that and I don't know if that was a deliberate decision.

I tried this and it still seems to work. The steps I ran through were:

1. Set "Background process limit" to "No background processes"
2. Open a form with an image widget
3. Click "Take picture"
4. Take a picture in the camera app

In this case in both v2023.1.2 and v2023.2.2 I return to Collect with the picture. Are you seeing different results? I had thought you were probably correct though and that the Collect process would be in "the background" while the Camera is doing it's thing. Maybe Android keeps processes waiting for results in the foreground? Maybe this differs between API levels if that's the case?

grzesiek2010 replied:
I think in my scenario there was one extra step which was switching to another app after opening the camera app. So it would be:

1. Set "Background process limit" to "No background processes"
2. Open a form with an image widget
3. Click "Take picture"
4. Switch to another app and get back to the camera app
5. Take a picture in the camera app

seadowg replied:
Ah I see. That flow is definitely not working any longer (it is in v2022.4.4). I can switch this PR from removing that code, to getting it working again!

seadowg committed:
Make sure process restore is simulated correctly

seadowg committed:
Spike out being able to return external data after a process restore

seadowg committed:
Stop restores loading first question instead of hierarchy if it was previously up

seadowg commented:
@getodk/testers I'm marking this as "needs testing" early as I'm currently not able to write tests for the fix for the issue or for the new one @grzesiek2010 found:

1. Set "Background process limit" to "No background processes"
2. Open a form with an image widget
3. Click "Take picture"
4. Switch to another app and get back to the camera app
5. Take a picture in the camera app

Please verify that both problems are fixed and also have a go at finding any other related problems around process restores and "external" data. I can look at working out automated testing once we know my fixes actually work!

seadowg committed:
Add Robolectric implementation of process restore simulation

seadowg committed:
Replace instrumentation with local test

seadowg committed:
Duplicate class and remove commonTest

Sharing source sets this way is no longer supported:
https://issuetracker.google.com/issues/232007221#comment17

seadowg committed:
Move test forms to dedicated module

seadowg committed:
Integrate test file utils into main set

seadowg committed:
Use real form for local process restore test

seadowg committed:
Correct recreating simulation

seadowg committed:
Replace process restore with DialogFragment instrumentation test with local one

dbemke commented:
What would be the expected result now for the issue #5646?

seadowg commented:
> What would be the expected result now for the issue https://github.com/getodk/collect/issues/5646?

I think ideally it should return to the hierarchy in the same place (same group) as it was in before. Do you think that makes sense? Is that currently how it's behaving?

seadowg committed:
Migrate FormLoaderTask to use Scheduler

seadowg committed:
Use paused scheduler for FormFillingActiivtyTest

dbemke commented:
Yes, it returns to the hierarchy view - for me looks great as it's consistent with other cases of returning to a form

seadowg committed:
Make tests more specific

seadowg committed:
Pull out common helpers

seadowg committed:
Add test for returning from hierarchy

seadowg committed:
Pull out helpers for Espresso

seadowg committed:
Remove shadows from FormFillingAcitivity

seadowg committed:
Move extension to shared module

seadowg committed:
Pull out shared Espresso helpers

seadowg committed:
Add Espresso helper for asserting on intents

seadowg committed:
Remove unused assertion

seadowg committed:
Improve intent assertion

seadowg committed:
Add test for restoring and returning data

seadowg commented:
@dbemke does it look like the issue that @grzesiek2010 reported is fixed as well? It is as far as I can tell!

seadowg committed:
Correct lifecycle for process restore from external app

seadowg committed:
Ad helper for asserting on new intents

seadowg committed:
Add helper for restoring with result

seadowg committed:
Use receiveResult instead of protected method

seadowg commented:
I wrote a post going deeper into how all testing works here which might be useful for @grzesiek2010: https://www.seadowg.com/2023/08/14/simulating-activity-recreation-with-robolectric.html.

seadowg committed:
Avoid ActivityController mem leaks

dbemke commented:
> does it look like the issue that @grzesiek2010 reported is fixed as well?

Yes. The issue is fixed

dbemke commented:
I don't know if the PR should influence also this case. In video widget minimizing the app while recording a video with no background processes settings, after going back the recording is stopped/disappears (no crash and possible to start a new recording, but the recording which was on is not saved).

seadowg commented:
> I don't know if the PR should influence also this case. In video widget minimizing the app while recording a video with no background processes settings, after going back the recording is stopped/disappears (no crash and possible to start a new recording, but the recording which was on is not saved).

I wouldn't expect it to fix this. Does that issue also occur on `master`/Play Store version?

dbemke commented:
On master version after minimizing it's possible to start a new recording but after tapping "V" to confirm to upload the video, the user is moved to the first page of the form and none of the recordings is saved. In the PR the second recording (the one started after minimizing) is saved.

seadowg commented:
> On master version after minimizing it's possible to start a new recording but after tapping "V" to confirm to upload the video, the user is moved to the first page of the form and none of the recordings is saved. In the PR the second recording (the one started after minimizing) is saved.

So it's changed, but it's (arguably) better now? Sounds like there is still a bug around not being able to record a a video in the background though that we should file separately?

dbemke commented:
> So it's changed, but it's (arguably) better now? Sounds like there is still a bug around not being able to record a a video in the background though that we should file separately?

I think it's better. I'll file the issue after merging this PR.

dbemke commented:
We finished testing for now so we'll remove the label "needs testing". As discussed in Slack, please let us know when/if we should continue testing in case of some changes.
Issues mentioned in the PR are fixed. The issue connected with video widget will be filed separately.

grzesiek2010 commented on test-forms/build.gradle.kts:
```diff
@@ -0,0 +1,9 @@
+}
+
+java {
+    sourceCompatibility = JavaVersion.VERSION_1_7
```
We have recently started using java 17 in other modules so here it should be the same not 7

grzesiek2010 commented on collect_app/src/androidTest/java/org/odk/collect/android/support/rules/FormEntryActivityTestRule.kt:
```diff
@@ -90,9 +90,20 @@ class FormEntryActivityTestRule : ExternalResource() {
+        return this
+    }
+
+    fun simulateProcessRestart(): FormEntryActivityTestRule {
```
I imagine that we might need those methods in other tests too. Wouldn't it be better to put them in `CollectTestRule` or a helper class so that they can be easily used in tests that start with creating a new project?

seadowg replied:
Ah there's actually a bunch of clean up I should now that we've moved that process restore stuff to Robolectric.

grzesiek2010 commented on collect_app/src/androidTest/java/org/odk/collect/android/support/rules/FormEntryActivityTestRule.kt:
```diff
@@ -90,9 +90,20 @@ class FormEntryActivityTestRule : ExternalResource() {
     }
 
-    fun restoreActivity() {
+    fun restoreActivity(): FormEntryActivityTestRule {
```
This method is no longer used.

grzesiek2010 commented on collect_app/src/androidTest/java/org/odk/collect/android/TestSettingsProvider.kt:
```diff
@@ -0,0 +1,21 @@
+import org.odk.collect.shared.settings.Settings
+
+// Use just for testing
+object TestSettingsProvider {
```
We have a separate class for android and unit tests. Does that mean it can't be shared?

seadowg replied:
The old way of sharing (using a shared source set) isn't supported anymore. The `androidTest` one is actually only used in 3 places, so I can just replace those calls so we don't have the duplicated object.

grzesiek2010 commented on collect_app/build.gradle:
```diff
@@ -389,6 +381,7 @@ dependencies {
     testImplementation Dependencies.androidx_test_core_ktx
 
     androidTestImplementation project(':androidtest')
+    androidTestImplementation project(':test-forms')
```
Should we move other resources like media files there too?

seadowg replied:
Yeah sounds good! I'll keep the instances and dbs in where they are for now, but move all the media files so everything for the test forms is in one place.

grzesiek2010 commented on testshared/src/main/java/org/odk/collect/testshared/EspressoHelpers.kt:
```diff
@@ -0,0 +1,45 @@
+        onView(withContentDescription(string)).perform(click())
+    }
+
+    fun assertIntents(vararg activityClasses: KClass<*>) {
```
This method is unused.

seadowg committed:
Fix java version

seadowg committed:
Remove unused method

seadowg committed:
Remove no longer needed test dummy

seadowg committed:
Inline settings setup in androidTest

seadowg committed:
Remove unused method

seadowg committed:
Move form media files to test-forms module
